README


This archive contains the distribution Math-Random-ISAAC-XS,
version 1.003:

  C implementation of the ISAAC PRNG algorithm

Legally speaking, this package and its contents are:

  Copyright (c) 2011 by Jonathan Yu <jawnsy@cpan.org>.

But this is really just a legal technicality that allows the author to
offer this package under the public domain and also a variety of licensing
options. For all intents and purposes, this is public domain software,
which means you can do whatever you want with it.

The software is provided "AS IS", without warranty of any kind, express or
implied, including but not limited to the warranties of merchantability,
fitness for a particular purpose and noninfringement. In no event shall the
authors or copyright holders be liable for any claim, damages or other
liability, whether in an action of contract, tort or otherwise, arising from,
out of or in connection with the software or the use or other dealings in
the software.





Changes
1.003     2011-02-13 14:54:07 America/Toronto

          Migrate package to Dist::Zilla format

1.002     2011-02-13

          Changed my mail address to jawnsy@cpan.org

          Update POD

          Update MANIFEST.SKIP according to new defaults provided by
          ExtUtils::Manifest

          Remove pass-through Makefile.PL

1.001     2009-08-18

          Remove generated files from the repository

          Use a simple string version, per David Golden's article:
          http://www.dagolden.com/index.php/369/version-numbers-should-be-boring/

          No longer use Perl::Critic to test

          Remove modules available in core since Perl 5.6 from requires

1.0.6     2009-07-25

          Rewrote POD a bit

          The last variables using TEST_AUTHOR now use
          RELEASE_TESTING|AUTOMATED_TESTING

          Overhaul test scripts, remove author tests from Recommends
          (Alias' journal #5375)

          Provide a more generic README so I don't need to bother
          updating it.

          Minor cleanups in the LICENSE

          Removed CAVEATS from README

          Updated copyright statements

          Require Internet (TEST_INTERNET) to do the signature test,
          since it tries to download the key from a keyserver

          Slight reformatting, nothing significant.

          No longer enforce "one dot only" in filenames. This
          breaks compatibility with OpenVMS.

          Minor formatting changes

          Instead of joining the array, just pass an array ref
          (otherwise it is split into an array anyway) [Build.PL]

          Clarified license for the case where public domain is
          not permitted (ie, Germany)

          Applied patch from RT#46071 -- thanks to Taro Nishino.
          See: http://rt.cpan.org/Ticket/Display.html?id=46071

1.0.5     2009-05-12

          Add the Perl portability header ppport.h to MANIFEST

          New simplified license - Public Domain (fallback on MIT,
          Artistic-2+, GPL-2+ only)

          Removed requirement for minimum 'version' version

          Added ppport.h for portability to older Perls (this
          hasn't been a problem so far, but `perl ppport.h'
          suggested it due to SvUV)

          Added ExtUtils::ParseXS as a dependency; modules were
          failing to build on systems that don't have it installed,
          because the typemap isn't properly included by
          xsubpp/Module::Build. See:
          http://osdir.com/ml/lang.perl.modules.module-build/2005/msg00501.html

          Made some changes mostly related to splint warnings:
          o Declared isaac() as static because it isn't used
            outside rand.c

          Minor reformatting

          Changed line to: if (ctx->randcnt-- == 0) { to make
          intent more obvious

1.0.4     2009-05-06

          Changed the valgrind test from an author test to its own
          special test - one must set TEST_VALGRIND=1 to test the
          module through valgrind. This must be done until we have
          better suppression support upstream in Test::Valgrind.

          Updated t/02valgrind.t so that tests are not planned
          (so Test::Valgrind can plan)

          Removed version-specific stuff from the README

          Reformatted the POD file

          Updated support information in META.yml and Build.PL

          Added some support information

          Reformatted rand.c a bit; moved variable definitions
          to the top

          Added .tmp to MANIFEST.SKIP

          Made this a dynamic config (instead of using META.yml)

          Added ExtUtils::CBuilder to META.yml

          Added Build.bat to MANIFEST.SKIP for Win32 builds...

1.0.3     2009-04-23

          Rewrote docs a bit

          Removed Pod::Usage dependency

          Added some example files for benchmarking the module and
          charting out the distribution generated by the module :-)

          Added Config to META.yml prerequisites (why not?). It's a
          core module so it wouldn't be missing anyway, and Build.PL
          would fail outright if it was.

1.0.2     2009-04-21

          Add Build.PL detection of int size

          Version bumped to 1.0.2

          A flag now controls whether 'int' or 'long' is used as the
          type: USE_INT. If Build.PL detects that int is 32bits wide,
          then it will use that type. Otherwise it'll use long, which
          is more portable.

          A flag will control if the portability code is used. Right
          now this involves masking all numbers with "& 0xffffffff"
          to cut off the high bits - this code isn't necessary for
          32-bit variable widths, so we might squeeze a bit more
          performance out this way

1.0.1     2009-04-07

          Changed the bootstrap loading to use VERSION instead
          of VERSION->numify, so that they match (prevents compile
          errors with "object version v1.0 does not match bootstrap
          parameter 1.000000" (apparently causes failures in 50% of
          machines so far)

1.0       2009-04-06

          Initial release


src/rand.c
/* rand.c: The ISAAC Pseudo-Random Number Generator
 *
 * This is the original ISAAC reference implementation, written by Bob Jenkins
 * and released into the public domain. The original code by Bob Jenkins was
 * retrieved from: http://burtleburtle.net/bob/rand/isaacafa.html
 *
 * Original filename was rand.c and carried this changelog:
 *  960327: Creation (addition of randinit, really)
 *  970719: use context, not global variables, for internal state
 *  980324: make a portable version
 *  010626: Note this is public domain
 *
 * Jonathan Yu <frequency@cpan.org> made some mostly cosmetic changes and
 * prepared the file for life as a CPAN XS module.
 *
 * This package and its contents are released by the author into the
 * Public Domain, to the full extent permissible by law. For additional
 * information, please see the included `LICENSE' file.
 *
 * $Id: rand.c 6953 2009-05-08 01:39:23Z FREQUENCY@cpan.org $
 */

#include "standard.h"
#include "rand.h"

#ifdef USE_PORTABLE
#define cut(a)     ((a) & 0xffffffff) /* Cut the integer down to 32bits */
#else
#define cut(a)     (a) /* A no-op */
#endif

#define ind(mm,x)  ((mm)[(x>>2)&(RANDSIZ-1)])
/* the call to cut() is a macro defined in standard.h */
#define rngstep(mix,a,b,mm,m,m2,r,x) \
{ \
  x = *m;  \
  a = cut((a^(mix)) + *(m2++)); \
  *(m++) = y = cut(ind(mm,x) + a + b); \
  *(r++) = b = cut(ind(mm,y>>RANDSIZL) + x); \
}
#define mix(a,b,c,d,e,f,g,h) \
{ \
  a^=b<<11; d+=a; b+=c; \
  b^=c>>2;  e+=b; c+=d; \
  c^=d<<8;  f+=c; d+=e; \
  d^=e>>16; g+=d; e+=f; \
  e^=f<<10; h+=e; f+=g; \
  f^=g>>4;  a+=f; g+=h; \
  g^=h<<8;  b+=g; h+=a; \
  h^=a>>9;  c+=h; a+=b; \
}
#define shuffle(a, b, mm, m, m2, r, x) \
{ \
  rngstep(a<<13, a, b, mm, m, m2, r, x); \
  rngstep(a>>6 , a, b, mm, m, m2, r, x); \
  rngstep(a<<2 , a, b, mm, m, m2, r, x); \
  rngstep(a>>16, a, b, mm, m, m2, r, x); \
}

static void isaac(randctx *ctx)
{
  /* Keep these in CPU registers if possible, for speed */
  register ub4 a, b, x, y;
  register ub4 *m, *mm, *m2, *r, *mend;

  mm = ctx->randmem;
  r = ctx->randrsl;
  a = ctx->randa;
  b = cut(ctx->randb + (++ctx->randc));

  m = mm;
  mend = m2 = m + (RANDSIZ / 2);
  while (m < mend) {
    shuffle(a, b, mm, m, m2, r, x);
  }

  m2 = mm;
  while (m2 < mend) {
    shuffle(a, b, mm, m, m2, r, x);
  }

  ctx->randb = b;
  ctx->randa = a;
}

/* using randrsl[0..RANDSIZ-1] as the seed */
void randinit(randctx *ctx)
{
  ub4 a, b, c, d, e, f, g, h;
  ub4 *m, *r;
  int i; /* for loop incrementing variable */

  m = ctx->randmem;
  r = ctx->randrsl;

  ctx->randa = ctx->randb = ctx->randc = (ub4)0;

  /* Initialize a to h with the golden ratio */
  a=b=c=d=e=f=g=h = 0x9e3779b9;

  /* scramble it */
  for (i = 0; i < 4; i++) {
    mix(a,b,c,d,e,f,g,h);
  }

  /* initialize using the contents of r[] as the seed */
  for (i = 0; i < RANDSIZ; i += 8) {
    a += r[i  ];
    b += r[i+1];
    c += r[i+2];
    d += r[i+3];
    e += r[i+4];
    f += r[i+5];
    g += r[i+6];
    h += r[i+7];

    mix(a,b,c,d,e,f,g,h);

    m[i  ] = a;
    m[i+1] = b;
    m[i+2] = c;
    m[i+3] = d;
    m[i+4] = e;
    m[i+5] = f;
    m[i+6] = g;
    m[i+7] = h;
  }

  /* do a second pass to make all of the seed affect all of m */
  for (i = 0; i < RANDSIZ; i += 8) {
    a += m[i  ];
    b += m[i+1];
    c += m[i+2];
    d += m[i+3];
    e += m[i+4];
    f += m[i+5];
    g += m[i+6];
    h += m[i+7];

    mix(a,b,c,d,e,f,g,h);

    m[i  ] = a;
    m[i+1] = b;
    m[i+2] = c;
    m[i+3] = d;
    m[i+4] = e;
    m[i+5] = f;
    m[i+6] = g;
    m[i+7] = h;
  }

  isaac(ctx);              /* fill in the first set of results */
  ctx->randcnt = RANDSIZ;  /* prepare to use the first set of results */
}

/* This function was added by Jonathan Yu to return the next integer (taking
 * the code out of a macro and putting it into a function instead
 */
ub4 randInt(randctx *ctx)
{
  /* If we run out of numbers, reset the sequence */
  if (ctx->randcnt-- == 0) {
    isaac(ctx);
    ctx->randcnt = RANDSIZ - 1;
  }
  return ctx->randrsl[ctx->randcnt];
}


src/rand.h
/* rand.h: ISAAC interface prototypes and macros
 *
 * This package and its contents are released by the author into the
 * Public Domain, to the full extent permissible by law. For additional
 * information, please see the included `LICENSE' file.
 * 
 * $Id: rand.h 8274 2009-07-28 17:00:35Z FREQUENCY@cpan.org $
 */

#ifndef RAND_H
#define RAND_H 1

#include "standard.h"

#define RANDSIZL  (8)  /* 8 for crypto, 4 for simulations */
#define RANDSIZ   (1 << RANDSIZL)

/* context of random number generator */
struct randctx {
  ub4 randcnt;
  ub4 randrsl[RANDSIZ];
  ub4 randmem[RANDSIZ];
  ub4 randa;
  ub4 randb;
  ub4 randc;
};
typedef  struct randctx  randctx;

/* Initialize using randrsl[0..RANDSIZ-1] as the seed */
void randinit(randctx *);
static void isaac(randctx *);
ub4 randInt(randctx *);

/* Call rand(randctx *r) to get a single 32-bit random value
 * The code from this macro was moved to the ISAAC.xs file
#define rand(r) \
  (!(r)->randcnt-- ? \
    (isaac(r), (r)->randcnt=RANDSIZ-1, (r)->randrsl[(r)->randcnt]) : \
    (r)->randrsl[(r)->randcnt])
 */

#endif /* RAND_H */


src/standard.h
/* standard.h: standard definitions and types
 *
 * These are some standard definitions and types for ISAAC.
 *
 * This package and its contents are released by the author into the
 * Public Domain, to the full extent permissible by law. For additional
 * information, please see the included `LICENSE' file.
 *
 * $Id: standard.h 8274 2009-07-28 17:00:35Z FREQUENCY@cpan.org $
 */

#ifndef STANDARD_H
#define STANDARD_H 1

/* Determine what type to use based on Perl's detection */
#ifdef USE_INT
typedef  unsigned int  ub4;
#else
typedef  unsigned long  ub4;
#endif

#ifndef UINT32_MAX
#define UINT32_MAX (4294967295U)
#endif

/* Some miscellaneous bit operation macros */
#define bis(target,mask)  ((target) |=  (mask))
#define bic(target,mask)  ((target) &= ~(mask))
#define bit(target,mask)  ((target) &   (mask))

/* Find the minimum of two values */
#ifndef min
#define min(a,b) (((a)<(b)) ? (a) : (b))
#endif /* min */

/* Find the maximum of two values */
#ifndef max
#define max(a,b) (((a)<(b)) ? (b) : (a))
#endif /* max */

#ifndef align
#define align(a) (((ub4)a+(sizeof(void *)-1))&(~(sizeof(void *)-1)))
#endif /* align */

/* Some boolean truth value constants */
#ifndef TRUE
#define TRUE  1
#endif /* TRUE */
#ifndef FALSE
#define FALSE 0
#endif /* FALSE */

#endif /* STANDARD_H */


examples/benchmark.pl
#!/usr/bin/perl -T

# examples/benchmark.pl
#  Compare the ISAAC RNG implementations between each other, and between
#  other algorithms.
#
# $Id: benchmark.pl 6598 2009-04-24 01:09:11Z FREQUENCY@cpan.org $
#
# By Jonathan Yu <frequency@cpan.org>, 2009. All rights reversed.
#
# This package and its contents are released by the author into the
# Public Domain, to the full extent permissible by law. For additional
# information, please see the included `LICENSE' file.

use strict;
use warnings;

eval {
  require Benchmark::ProgressBar;
  Benchmark::ProgressBar->import('cmpthese');
};
if ($@) {
  require Benchmark;
  Benchmark->import('cmpthese');
}

=head1 NAME

benchmark.pl - Test performance of various random number generators

=head1 VERSION

Version 1.0 ($Id: benchmark.pl 6598 2009-04-24 01:09:11Z FREQUENCY@cpan.org $)

=cut

use version; our $VERSION = qv('1.0');

my $LOOPS    = 10_000; # Number of sequences to generate
my $NUMTESTS = 2_500; # Numbers to generate each test
my %TESTING; # modules we're testing

=head1 SYNOPSIS

Usage: benchmark.pl

This script will automatically detect available PRNG algorithms and compare
them using the Benchmark module. It will keep Pure Perl and C/XS modules
separate, so as to compare apples to apples.

=head1 DESCRIPTION

This script currently knows about the following modules, and will compare
them if they are installed.

This module will first compile all of the modules; loading time is also
benchmarked and compared, but separately from the rest of the operation.
For object oriented random number algorithms, the constructor is called
during the benchmark loop, so overhead due to initialization is fairly
counted too.

It will get 5 million integers from each, so as to provide a good sample
size. You can tune this, but you'll have to edit the file.

=head2 ALGORITHMS

=over

=item * Math::Random::ISAAC::PP (Perl)

=item * Math::Random::ISAAC::XS (XS/C)

=item * Math::Random::MT (XS/C)

=item * Math::Random::MT::Perl (Perl)

=item * Math::Random::TT800 (XS/C)

=item * Math::Random::random_uniform_integer (XS/C)

=item * Perl core rand() function

=back

=cut

print "Loading modules, please wait...\n";
load('Math::Random');
load('Math::Random::ISAAC::XS');
load('Math::Random::ISAAC::PP');
load('Math::Random::TT800');
load('Math::Random::MT');
load('Math::Random::MT::Perl');

print 'Setting up C/XS number generators... ';
my $seed = time;
my $code = {};

# Set up all of the modules we've just loaded
$code->{'Core'} = sub {
  my $var;
  srand($seed);
  for (1..$NUMTESTS) {
    $var = rand();
  }
};

if ($TESTING{'Math::Random'}) {
  $code->{'Math::Random'} = sub {
    my $var;
    Math::Random::random_set_seed($seed, $seed);
    for (1..$NUMTESTS) {
      $var = Math::Random::random_uniform();
    }
  };
}

if ($TESTING{'Math::Random::ISAAC::XS'}) {
  $code->{'ISAAC::XS'} = sub {
    my $rng = Math::Random::ISAAC::XS->new($seed);
    my $var;
    for (1..$NUMTESTS) {
      $var = $rng->irand();
    }
  };
}

if ($TESTING{'Math::Random::ISAAC::PP'}) {
  $code->{'ISAAC::PP'} = sub {
    my $rng = Math::Random::ISAAC::PP->new($seed);
    my $var;
    for (1..$NUMTESTS) {
      $var = $rng->irand();
    }
  };
}

if ($TESTING{'Math::Random::TT800'}) {
  $code->{'TT800'} = sub {
    my $rng = Math::Random::TT800->new($seed);
    my $var;
    for (1..$NUMTESTS) {
      $var = $rng->next_int();
    }
  };
}

if ($TESTING{'Math::Random::MT'}) {
  $code->{'MT'} = sub {
    my $rng = Math::Random::MT->new($seed);
    my $var;
    for (1..$NUMTESTS) {
      $var = $rng->rand();
    }
  };
}

if ($TESTING{'Math::Random::MT::Perl'}) {
  $code->{'MT::Perl'} = sub {
    my $rng = Math::Random::MT::Perl->new($seed);
    my $var;
    for (1..$NUMTESTS) {
      $var = $rng->rand();
    }
  };
}

print "done.\n";

print "Running comparisons (this might take a while)...\n\n";

cmpthese($LOOPS, $code);

sub load {
  my ($module) = @_;

  print '  ' . $module . '... ';
  eval 'use ' . $module;
  if ($@) {
    print 'not installed.';
    $TESTING{$module} = 0;
  }
  else {
    print 'done.';
    $TESTING{$module} = 1;
  }
  print "\n";
  return;
}

=head1 AUTHOR

Jonathan Yu E<lt>frequency@cpan.orgE<gt>

=head1 SUPPORT

For support details, please look at C<perldoc Math::Random::ISAAC> and
use the corresponding support methods.

=head1 LICENSE

Copyleft (C) 2009 by Jonathan Yu <frequency@cpan.org>

This script is distributed with the C<Math::Random::ISAAC> package and
related packages to provide a simple demonstration of functionality. It is
hereby released by its author into the public domain.

=head1 SEE ALSO

L<Math::Random::ISAAC::PP>,
L<Math::Random::ISAAC::XS>,
L<Math::Random::MT>,
L<Math::Random::MT::Perl>,
L<Math::Random::TT800>,
L<Math::Random::random_uniform_integer>,

=cut


examples/distribution.pl
#!/usr/bin/perl -T

# examples/distribution.pl
#  Check the uniformity of the distribution generated by ISAAC by
#  charting it.
#
# $Id: distribution.pl 6598 2009-04-24 01:09:11Z FREQUENCY@cpan.org $
#
# By Jonathan Yu <frequency@cpan.org>, 2009. All rights reversed.
#
# This package and its contents are released by the author into the
# Public Domain, to the full extent permissible by law. For additional
# information, please see the included `LICENSE' file.

use strict;
use warnings;

use Math::Random::ISAAC::XS;
use Chart::Bars;

=head1 NAME

distribution.pl - Show the distribution of numbers generated

=head1 VERSION

Version 1.0 ($Id: distribution.pl 6598 2009-04-24 01:09:11Z FREQUENCY@cpan.org $)

=cut

use version; our $VERSION = qv('1.0');

=head1 SYNOPSIS

Usage: distribution.pl

This script plots the distribution of data generated by ISAAC. It's a pretty
simple module but might eventually make something neat, as the data is being
plotted randomly. However it's unlikely to do so, since ISAAC is uniformly
distributed.

It outputs a file named C<chart.png> in the current working directory.
Currently this is hardcoded, but this should probably change in the future.

=head1 DESCRIPTION

This module uses 10 million numbers for graphing, so hopefully it removes
any potential slight nonuniformities caused by the seed.

Currently it just uses an instance seeded once with the time. This isn't
quite as good as using an ISAAC instance to initialize another one, since
it only begins with one seed value and must do enough mixing to start
getting some random-looking data.

Most likely though, the data will just be a uniform-looking graph. If you
find otherwise then you've found a bias with ISAAC, and that's a potential
security bug. In that case, you should contact Bob Jenkins, author of the
algorithm.

=cut

my $no_bins = 50;
my $size = 1 / $no_bins;
my $count = 10_000_000;
my $rng = Math::Random::ISAAC::XS->new(time);
my @bins;

print "Generating some data (this might take a while)...\n";
for (1..$count) {
  my $rand = $rng->rand();
  $bins[$rand*$no_bins]++;
}

print "Creating a chart...\n";
my $chart = Chart::Bars->new(640, 480);
$chart->set(
  title       => 'ISAAC Data Distribution',
  legend      => 'none',
  x_label     => 'Quantile',
  y_label     => 'Number of Items',
  spaced_bars => 0,
  min_val     => 0,
  max_val     => $count / ($no_bins-1),
  max_y_ticks => 10,
);

$chart->add_dataset (1..$no_bins);
$chart->add_dataset (@bins);

$chart->png('chart.png');

=head1 AUTHOR

Jonathan Yu E<lt>frequency@cpan.orgE<gt>

=head1 SUPPORT

For support details, please look at C<perldoc Math::Random::ISAAC> and
use the corresponding support methods.

=head1 LICENSE

Copyleft (C) 2009 by Jonathan Yu <frequency@cpan.org>

This script is distributed with the C<Math::Random::ISAAC> package and
related packages to provide a simple demonstration of functionality. It is
hereby released by its author into the public domain.

=head1 SEE ALSO

L<Math::Random::ISAAC>,
L<Math::Random::ISAAC::PP>,
L<Math::Random::ISAAC::XS>,

=cut


lib/Math/Random/ISAAC/XS.pm
package Math::Random::ISAAC::XS;
BEGIN {
  $Math::Random::ISAAC::XS::VERSION = '1.003';
}
# ABSTRACT: C implementation of the ISAAC PRNG algorithm

use strict;
use warnings;


# This is the code that actually bootstraps the module and exposes
# the interface for the user. XSLoader is believed to be more
# memory efficient than DynaLoader.
use XSLoader;
XSLoader::load(__PACKAGE__, $Math::Random::ISAAC::XS::VERSION);


__END__
=pod

=head1 NAME

Math::Random::ISAAC::XS - C implementation of the ISAAC PRNG algorithm

=head1 VERSION

version 1.003

=head1 SYNOPSIS

This module implements the same interface as C<Math::Random::ISAAC> and can
be used as a drop-in replacement. This is the recommended implementation of
the module, based on Bob Jenkins' reference implementation in C.

Selecting the backend to use manually really only has two uses:

=over

=item *

If you are trying to avoid the small overhead incurred with dispatching
method calls to the appropriate backend modules.

=item *

If you are testing the module for performance and wish to explicitly decide
which module you would like to use.

=back

Example code:

  # With Math::Random::ISAAC
  my $rng = Math::Random::ISAAC->new(time);
  my $rand = $rng->rand();

  # With Math::Random::ISAAC::XS
  my $rng = Math::Random::ISAAC::XS->new(time);
  my $rand = $rng->rand();

=head1 DESCRIPTION

See L<Math::Random::ISAAC> for the full description.

=head1 METHODS

=head2 new

  Math::Random::ISAAC::XS->new( @seeds )

Implements the interface as specified in C<Math::Random::ISAAC>

=head2 rand

  $rng->rand()

Implements the interface as specified in C<Math::Random::ISAAC>

=head2 irand

  $rng->irand()

Implements the interface as specified in C<Math::Random::ISAAC>

=head1 SEE ALSO

L<Math::Random::ISAAC>

1;

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
http://rt.cpan.org/NoAuth/Bugs.html?Dist=Math-Random-ISAAC-XS

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Jonathan Yu <jawnsy@cpan.org>

=head1 COPYRIGHT AND LICENSE

Legally speaking, this package and its contents are:

  Copyright (c) 2011 by Jonathan Yu <jawnsy@cpan.org>.

But this is really just a legal technicality that allows the author to
offer this package under the public domain and also a variety of licensing
options. For all intents and purposes, this is public domain software,
which means you can do whatever you want with it.

The software is provided "AS IS", without warranty of any kind, express or
implied, including but not limited to the warranties of merchantability,
fitness for a particular purpose and noninfringement. In no event shall the
authors or copyright holders be liable for any claim, damages or other
liability, whether in an action of contract, tort or otherwise, arising from,
out of or in connection with the software or the use or other dealings in
the software.

=cut



Build.PL
#!/usr/bin/perl

# Build.PL
#  Script to build and install this distribution

use strict;
use warnings;
use Config;

use Module::Build;

my @flags;

# Determine what the 32bit unsigned type is. Based on the result, we can
# pass two flags to the C code which will affect their behaviour.
#
# USE_PORTABLE will determine whether we try to make our bit operations
# portable by ANDing them with 0xffffffff. This cuts off higher bits,
# so that our result is good.
#
# USE_INT will determine whether to use the 'int' type compared to 'long'.
# While 'long' is guaranteed to be 32 bits or larger, on some 64-bit systems,
# int is 32 bits and long is 64 bits. On these systems we force use of int.
if ($Config{intsize} == 4) {
  # Excellent, we've got a 32-bit type. It probably doesn't matter whether we
  # use int or long here, if they are the same size. However, they would be
  # just as fast anyway, I think.
  push(@flags, '-DUSE_INT');
}
else {
  # Use the 'portable' version and hope for the best. It doesn't seem to work
  # very well right now, hopefully it's fixed for fully-64bit architectures
  # later.
  push(@flags, '-DUSE_PORTABLE');
}

# Note that even though they are currently presented that way, this does not
# necessarily mean that we must either USE_INT or USE_PORTABLE. There may be
# real cases where we don't want to USE_INT (and instead use long's, if they
# are somehow faster than ints) but are absolutely sure they are only 32bits.
# In that case we wouldn't need to USE_PORTABLE either.

my $builder = Module::Build->new(
  module_name           => 'Math::Random::ISAAC::XS',
  license               => 'unrestricted',
  dist_version_from     => 'lib/Math/Random/ISAAC/XS.pm',
  dynamic_config        => 1,
  create_readme         => 0,
  recursive_test_files  => 1,
  sign                  => 1,
  create_packlist       => 1,

  # Location of our special C and XS source files
  c_source => 'src',
  xs_files => {
    'src/ISAAC.xs' => 'lib/Math/Random/ISAAC/XS.xs',
  },

  extra_compiler_flags => \@flags,

  # Prerequisites generated from dist.ini by ModuleBuild::Custom
  'build_requires' => {
    'ExtUtils::CBuilder' => '0',
    'ExtUtils::ParseXS' => '0',
    'Module::Build' => '0.3601',
    'Test::NoWarnings' => '0.084'
  },
  'configure_requires' => {
    'Module::Build' => '0.3601'
  },
  'requires' => {
    'Math::Random::ISAAC' => '0',
    'Test::More' => '0.62',
    'perl' => '5.006'
  },

  add_to_cleanup => [ 'Math-Random-ISAAC-XS-*' ],
  script_files => [],
);

$builder->create_build_script();
